File: android/src/main/kotlin/com/leonardo/have_app/ApplicationService.kt
package com.leonardo.have_app

import com.leonardo.have_app.core.types.GetApplication

interface ApplicationService {
    /**
     * Get application by [packageName]
     * @return if does application exists returns [Application],
     * does not return an [Exception]
     */
    fun getApplication(packageName: String): GetApplication
}

class PackageManagerService(private val applicationGateway: ApplicationGateway): ApplicationService {
    override fun getApplication(packageName: String): GetApplication {
        return applicationGateway.getApplication(packageName)
    }
}

File: lib/domain/services/application_service.dart
import 'package:dartz/dartz.dart';
import 'package:have_app/core/faults/application_exception.dart';
import 'package:have_app/data/gateways/application_gateway.dart';
import 'package:have_app/data/models/models.dart';
import 'package:have_app/domain/entities/entities.dart';

abstract class ApplicationService {
  const ApplicationService();

  /// Get application by [packageName] and if does application exists
  /// returns [ApplicationModel], does not return an [ApplicationException]
  Future<Either<ApplicationException, Application>> getPackage(
    GetApplication getApplication,
  );

  /// Get all installed applications in device
  Future<Either<ApplicationException, List<ApplicationModel>>> getAllInstalled();
}

class ApplicationServiceImpl implements ApplicationService {
  const ApplicationServiceImpl({required this.applicationGateway});

  final ApplicationGateway applicationGateway;

  @override
  Future<Either<ApplicationException, Application>> getPackage(
      GetApplication getApplication) {
    return applicationGateway.getPackage(
      GetApplicationModel(packageName: getApplication.packageName),
    );
  }

  @override
  Future<Either<ApplicationException, List<ApplicationModel>>> getAllInstalled() {
    return applicationGateway.getAllInstalled();
  }
}
